packages/strapi-plugin-upload/admin/src/containers/InputModalStepperProvider/reducer.js

```diff
@@ -0,0 +1,169 @@
+import produce from 'immer';
+import { intersectionWith, differenceWith, unionWith } from 'lodash';
+
+import createNewFilesToUploadArray from '../../utils/createNewFilesToUploadArray';
+
+const initialState = {
+  selectedFiles: [],
+  files: [],
+  filesToUpload: [],
+  fileToEdit: null,
+  tab: null,
+  params: {
+    _limit: 10,
+    _start: 0,
+    _q: '',
+    filters: [],
+  },
+  currentStep: 'list',
+};
+
+const reducer = (state, action) =>
+  // eslint-disable-next-line consistent-return
+  produce(state, draftState => {
+    switch (action.type) {
+      case 'ON_CHANGE': {
+        draftState.fileToEdit[action.keys.split('.')] = action.value;
+        break;
+      }
+      case 'GET_DATA_COUNT_SUCCEEDED': {
+        draftState.count = action.data.count;
+        break;
+      }
+      case 'GET_DATA_SUCCEEDED': {
+        draftState.files = action.data;
+        break;
+      }
+      case 'SET_PARAM': {
+        const { name, value } = action.param;
+
+        if (name === 'filters') {
+          draftState.params.filters.push(value);
+          break;
+        }
+
+        if (name === '_limit') {
+          draftState.params._start = 0;
+        }
+
+        draftState.params[name] = value;
+        break;
+      }
+      case 'ON_FILE_SELECTION': {
+        const { value, id } = action;
+
+        if (value) {
+          const fileToStore = state.files.find(file => file.id === id);
+          draftState.selectedFiles.push(fileToStore);
+          break;
+        }
+
+        const index = draftState.selectedFiles.findIndex(item => item.id === id);
+        draftState.selectedFiles.splice(index, 1);
+        break;
+      }
+      case 'TOGGLE_SELECT_ALL': {
+        const comparator = (first, second) => first.id === second.id;
+        const isSelected =
+          intersectionWith(state.selectedFiles, state.files, comparator).length ===
+          state.files.length;
+
+        if (isSelected) {
+          draftState.selectedFiles = differenceWith(state.selectedFiles, state.files, comparator);
+          break;
+        }
+
+        draftState.selectedFiles = unionWith(state.selectedFiles, state.files, comparator);
+        break;
+      }
+      case 'SET_FILE_ERROR': {
+        draftState.filesToUpload.forEach((fileToUpload, index) => {
```

Comment:
Maybe you can use a map here instead of a `forEach` I am not familiar with the lib though

Reply:
The difference between map and forEach is that the forEach function will mutate the current array. The basic idea with Immer is that you will apply all the changes to a temporary draftState (mutable object) and then, Immer will produce a new immutable one.

Reply:
There is no new API here just pure JS. The immer part is just the call of `produce(`.